Term: Rewrite Konsole font detection, closes #576

mstraube opened the pull request:
I've rewritten the font detection for Konsole. It now
- works for `org.kde.konsole` and `org.kde-konsole-$PPID`
- works if the profile filename differs from the profile name
- doesn't use hardcoded config file paths
- works for KDE5 and KDE4

Tested on Arch, Kubuntu, openSUSE, Slackware, KDE neon.

Note:
With Konsole versions affected by https://bugs.kde.org/show_bug.cgi?id=358083 (I think..)
it's possible that we have two profile files containing the same profile name.

I noticed this behavior (with a new created user) on Kubuntu 16.04:
Starting Konsole -> edit current profile -> close konsole -> restart it -> edit current profile
Produced two differnt profile files with the same profile name in it.

That's the only situation I was able to produce two different profile files containing the same profile name.
But it's possible. So we have to make sure we get a valid filename from the grep command. ( head -n 1 )

Issues:
Detecting the fallback font fails if no fixed font is defined in kdeglobals.
This is the case on fresh installations or for new created users for example.

Konsole internally uses `QFontDatabase::systemFont(QFontDatabase::FixedFont)`
or `KGlobalSettings::fixedFont()` in qt4 versions.

The font settings are written to kdeglobals once the user changes fonts in the KDE System-Settings GUI.

So if the user doesn't edit a profile or the system fonts, no font will be detected.
I'm afraid there is not much we can do..

Maybe someone else knows a way to get the fixed Qt system font without writing (e.g. C++ or python) small helper programs? I couldn't find any existing commands.

dylanaraps commented:
I really don't like how complex this has become. I think it's better if we remove Konsole support from `term_font()` entirely, at least until we come up with a simpler solution.

@konimex, @iandrewt, what do you think about this?

iandrewt commented:
If theres a way to simplify this then yeah, I'd go with that. But at 33 line additions just for one terminal emulator support, that's almost like `dbus` level horribleness

mstraube commented:
Yes, it's a bit complex. I looked at it again and removed fallback to kdeglobals (since it's useless if the user didn't edit the system font settings), removed kde4 support and switched back to the hardcoded default path for the profile files. 

I ended up with [this](https://gist.github.com/mstraube/074b1d278268925f528928977f656db3). Once the user has edited a profile and Konsole uses the default path it should work.

Happy new year! :)

dylanaraps commented:
I left some comments.

Happy new year. :)

mstraube commented:
Pushed the simplified version.

dylanaraps commented on neofetch:
```diff
@@ -1568,25 +1568,26 @@ get_term_font() {
+            # We could have two profile files for the same profile name, take first match
+            profile_filename="$(grep -l "Name=${profile}" "${HOME}"/.local/share/konsole/*.profile)"
+            profile_filename="${profile_filename/$'\n'*}"
+            [[ "$profile_filename" ]] && term_font="$(awk -F '=|,' '/Font=/ {print $2}' "${profile_filename}")"
```
Remove the `{}` from `$profile_filename`.

dylanaraps commented:
LGTM now :+1:

Is this ready to merge @mstraube?

mstraube commented:
Yes I think so, just one question. Should it also print the font size?
That would only require to change the print command to `{print $2 " " $3}`. I'm undecided...

mstraube commented:
Ok, updated.

I also thought about kde4 support. It could be done with a function similar to kde_config_dir() for the directory with the profiles. `kde_data_dir="$(kf5-config --path data)" `for example.

pro:
- support kde4
- don't use hardcoded path

con:
- needs extra function
- slower

But I guess there are not many distros, beside Slackware, that still have kde4?
For now I would leave it as it is.

dylanaraps commented:
:+1:
